[Doc] Refactor field (#6006)
taichi-dev/taichi: docs/lang/articles/basic/field.md

## Scalar fields

Scalar fields refer to the fields that store scalars and are the most basic fields. A 0D scalar field is a single scalar; a 1D scalar field is a 1D array of scalars; a 2D scalar field is a 2D array of scalars, and so on and so forth.

### Declaration

The simplest way to declare a scalar field is to call `ti.field(dtype, shape)`, where `dtype` is a primitive data type as explained in [type system](../type_system/type.md) and `shape` is a tuple of integers.

> + To declare a 0D scalar field, set its shape to the empty tuple `()`.

Reviewer: ```suggestion
+ To declare a 0D scalar field, set its shape to the empty tuple `()`:
```

    ```python
    # Declare a 0D scalar field whose data type is f32
    f_0d = ti.field(ti.f32, shape=())  # 0D field
    ```

    An illustration of `f_0d` is shown below:

    ```
        ┌─────┐
        │     │
        └─────┴
        └─────┘
     f_0d.shape=()
    ```

+ To declare a 1D scalar field of length `n`, set its shape to `n` or `(n,)`:

    ```python
    f_1d = ti.field(ti.i32, shape=9)  # 1D field
    ```

>     An illustration of `f_1d` shown is below:

Reviewer: ```suggestion
    An illustration of `f_1d` is shown below:
```

    ```
    ┌───┬───┬───┬───┬───┬───┬───┬───┬───┐
    │   │   │   │   │   │   │   │   │   │
    └───┴───┴───┴───┴───┴───┴───┴───┴───┘
    └───────────────────────────────────┘
            f_1d.shape = (9,)
    ```

> + To declare a 2D scalar field, set the size of its first two dimensions, i.e. the number of rows and columns, respectively. For example, the following code defines a 2D scalar field of shape (3, 6), which has 3 rows and 6 columns:

Reviewer: ```suggestion
+ To declare a 2D scalar field, set the sizes of its first two dimensions, i.e., the number of rows and columns, respectively. For example, the following code defines a 2D scalar field of shape (3, 6), which has 3 rows and 6 columns:
```

+ To declare a 0D scalar field, set its shape to the empty tuple `()`.

    ```python
    # Declare a 0D scalar field whose data type is f32
    f_0d = ti.field(ti.f32, shape=())  # 0D field
    ```

    The memory layout of the field `f_0d` will be like:

    ```
        ┌─────┐
        │     │
        └─────┴
        └─────┘
     f_0d.shape=()
    ```

+ To declare a 1D scalar field of length `n`, set its shape to `n` or `(n,)`:

    ```python
    f_1d = ti.field(ti.i32, shape=9)  # 1D field
    ```

    The memory layout of the field `f_1d` will be like:

    ```
    ┌───┬───┬───┬───┬───┬───┬───┬───┬───┐
    │   │   │   │   │   │   │   │   │   │
    └───┴───┴───┴───┴───┴───┴───┴───┴───┘
    └───────────────────────────────────┘
            f_1d.shape = (9,)
    ```

+ To declare a 2D scalar field, set the size of its first two dimensions, i.e. the number of rows and columns, respectively. For example, the following code defines a 2D scalar field of shape (3, 6), which has 3 rows and 6 columns:

    ```python
    f_2d = ti.field(int, shape=(3, 6))  # 2D field
    ```

>     The memory layout of the field `f_2d` will be like:

Reviewer 2: ```suggestion
    An illustration of f_2d is below:
```
Let's avoid using `memory layout` here, which has a different meaning. Just use `An illustration of f_2d is below` is good enough.

                  ┌  ┌───┬───┬───┬───┬───┬───┐  ┐
                  │  │   │   │   │   │   │   │  │
                  │  ├───┼───┼───┼───┼───┼───┤  │
    f_2d.shape[0] │  │   │   │   │   │   │   │  │
        (=3)      │  ├───┼───┼───┼───┼───┼───┤  │
                  │  │   │   │   │   │   │   │  │
                  └  └───┴───┴───┴───┴───┴───┘  ┘
    ```

+ To declare a 3D scalar field, set the dimensions of the three axes:

  ```python
  f_3d = ti.field(ti.f32, shape=(32, 16, 8))  # 3D field
  ```

  You can think of this 3D scalar field as a 32&times;16 2D field whose elements are arrays of length 8:

  ```
                    f_3d.shape[1]
                        (=16)
                   ┌───────────────┐

                ┌  ┌───┬╌╌╌╌╌╌╌┬───┐  ┐
                │  │ * │       │   │  │
                │  ├───┼╌╌╌╌╌╌╌┼───┤  │
  f_3d.shape[0] │  ╎   ╎       ╎   ╎  │
      (=32)     │  ├───┼╌╌╌╌╌╌╌┼───┤  │
                │  │   │       │   │  │
                └  └───┴╌╌╌╌╌╌╌┴───┘  ┘

      ┌───┬───┬───┬───┬───┬───┬───┬───┐
  * = │   │   │   │   │   │   │   │   │
      └───┴───┴───┴───┴───┴───┴───┴───┘
      └───────────────────────────────┘
             f_1d.shape[2] = 8

>   ```

Reviewer 2: I feel like this is too verbose. The readers should understand what a high-dimensional array is. Let's remove the 3D case here.

Scalar fields of higher dimensions can be similarily defined.

:::warning

Taichi only supports fields of dimensions <= 8.

:::

### Access elements in a scalar field

> Once a field is declared, Taichi automatically assigns an initial value of zero to its elements. But we do not recommend using this undefined behavior.

Reviewer 2: ```suggestion
Once a field is declared, Taichi automatically assigns an initial value of zero to its elements.
```

Zero-initialization is defined behavior.

```python
x = ti.field(int, shape=(5, 5))
x.fill(1)  # set all elements in x to 1

@ti.kernel
def test():
    x.fill(-1)  # set all elements in x to -1
```

### Metadata

Metadata provides the basic information of a scalar field. You can retrieve the data type and shape of a scalar field via its `shape` and `dtype` properties:

```python
f_1d.shape  # (9,)
f_3d.dtype  # f32
```

## Vector fields

As the name suggests, vector fields are the fields whose elements are vectors. What a vector represents depends on the scenario of your program. For example, a vector may stand for the (R, G, B) triple of a pixel, the position of a particle, or the gravitational field in space.

### Declaration

Declaring a vector field where each element is an `N`-dimensional vector is similar to the way you declare a scalar field, except that you need to call the function `ti.Vector.field` instead of `ti.field` and specify `N` as the first positional argument.

For example, the following code snippet declares a 2D field of 2D vectors:

```python
# Declare a 3x3 vector field comprising 2D vectors
f = ti.Vector.field(n=2, dtype=float, shape=(3, 3))
```

> The memory layout of `f` will be like:

Reviewer 2: Avoid the term `memory layout`.

- To access the velocity vector at a specific position of the volumetric field above:

  `volumetric_field[i, j, k]`

- To access the `l`-th component of the velocity vector:

  `volumetric_field[i, j, k][l]`

> - If the vectors in the field have dimension <=4, then you can also use vector swizzling indices `xyzw` or `rgba` to access the components of the vectors:

Author: "provided that the vector has no more than four dimensions:"

It should be "provided that the dimension of the vector is no more than four".

- To access the velocity vector at a specific position of the volumetric field above:

  `volumetric_field[i, j, k]`

- To access the `l`-th component of the velocity vector:

  `volumetric_field[i, j, k][l]`

- Alternatively, you can use swizzling with the indices `xyzw` or `rgba` to access the components of a vector, provided that the dimension of the vector is no more than four:

  ```python
  volumetric_field[i, j, k].x = 1  # equivalent to volumetric_field[i, j, k][0] = 1
  volumetric_field[i, j, k].y = 2  # equivalent to volumetric_field[i, j, k][1] = 2
  volumetric_field[i, j, k].z = 3  # equivalent to volumetric_field[i, j, k][2] = 3
  volumetric_field[i, j, k].w = 4  # equivalent to volumetric_field[i, j, k][3] = 4
  volumetric_field[i, j, k].xyz = 1, 2, 3  # assign 1, 2, 3 to the first three components

>   volumetric_field[i, j, k].rgb = 1, 2, 3  # ibid.

Reviewer 2: Avoid `ibid`. Not clear what that means.

```python
# Declare a 1D struct field using the ti.Struct.field() method
particle_field = ti.Struct.field({
    "pos": ti.math.vec3,
    "vel": ti.math.vec3,
    "acc": ti.math.vec3),
    "mass": float,
  }, shape=(n,))
```

Alternatively, instead of *directly* using `ti.Struct.field()`, you can first declare a compound type `particle` and then create a field of it:

```python
# vec3 is a built-in vector type suppied in the `taichi.math` module
vec3 = ti.math.vec3
# Declare a struct composed of three vectors and one floating-point number
particle = ti.types.struct(
  pos=vec3, vel=vec3, acc=vec3, mass=float,
)
# Declare a 1D field of the struct particle using field()
particle_field = particle.field(shape=(n,))
```

### Access elements in a struct field

> You can access a member of an element in a struct field in either of the following ways: "index-first"  or "name-first".

Reviewer: ```suggestion
You can access a member of an element in a struct field in either of the following ways: "index-first" or "name-first".
```

```python
# vec3 is a built-in vector type suppied in the `taichi.math` module
vec3 = ti.math.vec3
# Declare a struct composed of three vectors and one floating-point number
particle = ti.types.struct(
  pos=vec3, vel=vec3, acc=vec3, mass=float,
)
# Declare a 1D field of the struct particle using field()
particle_field = particle.field(shape=(n,))
```

### Access elements in a struct field

There are two ways to access a member in a struct field, either by the "index-to-name" way or by the "name-to-index" way.

+ By the index-to-name way, you first access the struct by its index in the field, and then access the member by its name in the struct:

```python
# Set the position of the first particle in the field to origin [0.0, 0.0, 0.0]
particle_field[0].pos = vec3(0) # pos is a 3D vector
```

+ By the name-to-index way, as shown in the code below, you first access the field which gathers all the `mass` members in the struct field, and then use the index to access a specific one:

```python
particle_field.mass[0] = 1.0  # Set the mass of the first particle in the field to 1.0
```

> Since `paticle_field.mass` is a field that consists of all the `mass` members of the structs in `paticle_field`, we can also call its `fill()` method to set them to a specific value all at once:

Reviewer: ```suggestion
Considering that `paticle_field.mass` is a field consisting of all the `mass` members of the structs in `paticle_field`, we can also call its `fill()` method to set the members to a specific value all at once:
```
